liquidcore/config/views.py

```diff
@@ -231,3 +236,124 @@ def post(self, request, format=None):
+        permission_classes=[IsAdminUser],
+        url_name='generate'
+    )
+    def generate(sef, request):
+        serializer = serializers.VPNClientKeyLabelSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        label = serializer.validated_data['label']
+        client_key = VPNClientKey.objects.create(label=label)
+        serialized_client_key = serializers.VPNClientKeySerializer(client_key)
+        reconfigure_system()
+        return Response(data=serialized_client_key.data)
+
+    @detail_route(
+        methods=['post'],
+        permission_classes=[IsAdminUser],
+        url_name='revoke'
+    )
+    def revoke(self, request, id=None):
+        serializer = serializers.VPNClientKeyRevokeSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        client_key = self.get_object()
+        client_key.revoked = True
+        client_key.revoked_reason = serializer.validated_data['revoked_reason']
+        client_key.revoked_at = timezone.now()
+        client_key.revoked_by = request.user
+        client_key.save()
+
+        reconfigure_system()
+        return Response()
+
+    @detail_route(
+        methods=['get'],
+        permission_classes=[IsAdminUser],
+        url_name='download'
+    )
+    def download(self, request, id=None):
+        # TODO get client key from system call
+        ovpn_content = "dummy key here, don't mind me\n"
+        filename = 'client-key-{}.ovpn'.format(id)
+        response = Response(ovpn_content, content_type=OVPN_CONTENT_TYPE)
+        response['Content-Disposition'] = 'attachment; filename="{}"'.format(filename)
+        return response
+
+@api_view(['GET'])
+@permission_classes([IsAdminUser])
+def vpn_status(request):
+    def get_description(state):
+        if state['error_message']:
+            return 'Error.'
+        if not state['is_enabled']:
+            return "Disabled."
+        if state['is_running']:
+            return 'Up and running.'
+        else:
+            return 'Starting...'
+
+    vpn_setting = Setting.objects.get(name='vpn')
+    status = vpn_setting.data
+    # TODO get these with system calls
+    status['client']['is_running'] = False
+    status['server']['is_running'] = False
+    status['server']['registered_key_count'] = VPNClientKey.objects.count()
+    status['server']['active_connection_count'] = 0
+
+    status['client']['state_description'] = get_description(status['client'])
+    status['server']['state_description'] = get_description(status['server'])
+    return Response(status)
+
+@api_view(['PUT'])
+@permission_classes([IsAdminUser])
+def vpn_server_enabled(request):
+    serializer = serializers.IsEnabledSerializer(data=request.data)
+    serializer.is_valid(raise_exception=True)
+    enabled = serializer.validated_data['is_enabled']
+    vpn_setting = Setting.objects.get(name='vpn')
+    vpn = vpn_setting.data
+    vpn['server']['is_enabled'] = enabled
+    vpn_setting.data = vpn
```

Comment:
I get that you have to trigger `setattr` for `data`. If you don't like the back and forth, try a context manager:

```python
with vpn_setting.change_data() as data:
    data['server']['is_enabled'] = enabled
```

And let the context manager overwrite the `data` attribute when you're done.

Reply:
Yeah, that makes sense. But won't lose time on it now.